x-pack/solutions/security/plugins/security_solution/server/lib/detection_engine/rule_types/esql/esql.ts

```diff
@@ -70,42 +70,48 @@ export const esqlExecutor = async ({
   } = sharedParams;
   const loggedRequests: RulePreviewLoggedRequest[] = [];
   const ruleParams = completeRule.ruleParams;
-  /**
-   * ES|QL returns results as a single page. max size of 10,000
-   * while we try increase size of the request to catch all alerts that might been deduplicated
-   * we don't want to overload ES/Kibana with large responses
-   */
-  const ESQL_PAGE_SIZE_CIRCUIT_BREAKER = tuple.maxSignals * 3;
   const isLoggedRequestsEnabled = state?.isLoggedRequestsEnabled ?? false;
 
   return withSecuritySpan('esqlExecutor', async () => {
     const result = createSearchAfterReturnType();
-    let size = tuple.maxSignals;
     const dataTiersFilters = await getDataTierFilter({
       uiSettingsClient: services.uiSettingsClient,
     });
 
+    /**
+     * ES|QL returns results as a single page, max size of 10,000
+     * To mitigate this, we will use the maxSignals as a page size
+     * Wll keep track of the earlier found document ids and will exclude them in subsequent requests
+     * to avoid duplicates.
+     * This is a workaround until pagination is supported in ES|QL
+     * Since aggregating queries do not produce event ids, we will not exclude them.
+     * All alerts for aggregating queries are unique anyway
+     */
+    const excludedDocumentIds: string[] = [];
+    let iteration = 0;
     try {
-      while (
-        result.createdSignalsCount <= tuple.maxSignals &&
-        size <= ESQL_PAGE_SIZE_CIRCUIT_BREAKER
-      ) {
+      while (result.createdSignalsCount <= tuple.maxSignals) {
```

Comment:
I did not set a limit for how many iterations(pages) we would query.
It is in line with the rest of rules that do paginations(IM, query, new_terms)

The only limit, is https://www.elastic.co/docs/reference/query-languages/query-dsl/query-dsl-terms-query#terms-top-level-params max number of terms in query(65k). I don't think it would be possible to reach out this limit before rule timeouts. Since it's basically requires 65k alerts to be deduplicated within rule lookback time

Reply:
we still need to figure out what the limit on the size of the ids array is and handle that case. `terms` supports 65536 values in the array, ids may be the same and 65536 could be hit without using that many requests if the user sets max_signals to a high value, and we've often seen the lookback configured to large values relative to the rule interval.

I think we probably want a rule warning, similar to the max_signals warning, rather than an error resulting from a search failure if they hit the max number of excluded IDs.

Reply:
But if we encounter search error we won't be able to proceed with search further, that may result in missing alerts.
Should not we in this case stick to error and rule failure rather than warning?

Reply:
Limit seems to be different for ids.values.
Rule was running with queries including 300k ids in it.
I also asked in es-team: https://elastic.slack.com/archives/C9BA8D0P4/p1745835667868339

Reply:
> Should not we in this case stick to error and rule failure rather than warning?

I think it's similar enough to the max_signals circuit breaker that we should use the same log level for both. Rule failures have usually been reserved for cases where the rule does not run correctly, but in this case the rule runs as we expect - it just may not have found all alerts due to the rule/data configuration.